cluster/images/etcd/migrate/data_dir.go

```diff
@@ -142,8 +146,18 @@ func (v *VersionFile) Read() (*EtcdVersionPair, error) {
 
 // Write creates or overwrites the contents of the version.txt file with the given EtcdVersionPair.
 func (v *VersionFile) Write(vp *EtcdVersionPair) error {
-	data := []byte(fmt.Sprintf("%s/%s", vp.version, vp.storageVersion))
-	return ioutil.WriteFile(v.path, data, 0666)
+	pvp, err := v.Read()
+	if err != nil {
+		return fmt.Errorf("failed to read old version file %s: %v", v.path, err)
```

Comment:
Is this correct? Don't we write the file for the first time via this path too?

Reply:
I think we should split this PR into two:
- the actual bug fix (that always writes next-file and renames)
- the optimization to write it only if its different

Reply:
> Is this correct? Don't we write the file for the first time via this path too?

It is not. Thanks for catching.

> I think we should split this PR into two:
> - the actual bug fix (that always writes next-file and renames)
> - the optimization to write it only if its different

Split seems reasonable, will keep first part in this PR and I'll submit rest in other.